### ISSUE TITLE
XmlParseError when parsing entities inside an attribute

### ISSUE THREAD

**Issue by corliss** [AngleSharp/AngleSharp#162]
Synced to the latest code. This fragment:

```
<div style="color:blue;foo:&quot;(50.5,100),(100,200)&quot;">...</div>
```

Note: `foo` is the name of a custom attribute, its actual name is not relevant.

causes an exception to be thrown from `XmlEntityToken.GetEntity()`. At this point, the `_value` variable seems malformed: `color:blue;foo:quot`

Is this a bug?

**Comment by corliss** [comment]
BTW, I'm switching to using AngleSharp's XML parsing because I need to handle entities beyond the default five defined by XML. AngleSharp's `XmlTokenizer` usually returns `XmlTokenType.Entity`. However in the case of attributes it wants to resolve them itself. Why?

Also you had mentioned previously that XML was not the focus. So wanted to confirm: at this point is the xml tokenization/parsing supported for real use? I'm currently using `XmlReader`, and it does not support custom entities (apart from DTDs), but apart from that it is very solid.

**Comment by FlorianRappl** [comment]
Hm, yes seems like a bug. `quot` is a known entity in XML (there aren't many). Besides that, however, please note that the XML parser really throws exceptions (conforming with W3C) and that in general  HTML documents will result in invalid XML.

The reason for resolving entities is obvious. The attribute's value is always a string and therefore needs to be resolved immediately [to construct that string]. There is no (accessible or hidden) model behind it.

**Comment by corliss** [comment]
Thanks.

Just so I understand - in an XML where DTD's are supported, are the character entities declared by the DTDs allowed in attributes? 

Assuming they are, what I'm looking for is a way of telling AngleSharp the entities that are valid for this xml document (without using a DTD.) 

That will make the AngleSharp XML parser work for all entities without the complexity of DTDs and validation. `XmlReader` requires DTD's to make this work, which is why I am using AngleSharp. As I mentioned earlier, `XmlTokenType.Entity` allows the caller to resolve entities except for attributes. Some sort of hook for attributes would complete the picture.

Adding such a hook isn't urgent however; currently only the basic entities are being used in attributes.

Cheers

**Comment by FlorianRappl** [comment]
DTD (i.e. a validating XML parser) was once integrated (during v0.3 times), but isn't right now and won't be in the future. 

DTD is another complex beast and the burden to support DTD fully would totally overshadow AngleSharp's HTML / CSS purpose. Therefore (and it is always announced that way) AngleSharp only provides a non-validating XML parser (capable of parsing any valid XML, but not capable of validating the document against a DTD). Hence there are no additional entities; only the ones defined in XML. One could, however, think about providing a service that is used for resolving entities.

**Comment by corliss** [comment]
To be clear, I also don't want DTDs, I think you may have misread. 

Just to recap, I'm using just the tokenizer, not the parser. This allows resolving custom entities when `XmlTokenType.Entity` is returned. However, it does not permit resolving entities in attributes.

I agree that a service to resolve entities would be simple and effective.

**Comment by FlorianRappl** [comment]
I just wanted to be clear about DTD in AngleSharp. Anyway, thanks for report. I was able to spot the issue and fix the bug. The fix will be released with v0.9.3.

**Comment by corliss** [comment]
Great, thanks!
